Framework/Algorithms/src/SortXAxis.cpp

```diff
@@ -179,51 +188,45 @@ void SortXAxis::copyXandDxToOutputWorkspace(
  * using clone because of how histograms are supported, for the Y Axis and the E
  * Axis.
  *
- * @param workspaceIndicies the sorted vector of indicies
+ * @param workspaceIndices the sorted vector of indicies
  * @param inputWorkspace the unsorted input workspaces
  * @param outputWorkspace the empty output workspace
  * @param specNum the spectrum number being copied into
- * @param isAProperHistogram whether or not it has been determined to be a valid
- * histogram earlier on.
  */
 void SortXAxis::copyYandEToOutputWorkspace(
-    std::vector<std::size_t> &workspaceIndicies,
+    std::vector<std::size_t> &workspaceIndices,
     const Mantid::API::MatrixWorkspace &inputWorkspace,
-    Mantid::API::MatrixWorkspace &outputWorkspace, unsigned int specNum,
-    bool isAProperHistogram) {
-  // If Histogram data find the biggest index value and remove it from
-  // workspaceIndicies
-  if (isAProperHistogram) {
-    auto lastIndexIt =
-        std::find(workspaceIndicies.begin(), workspaceIndicies.end(),
-                  inputWorkspace.y(specNum).size());
-    workspaceIndicies.erase(lastIndexIt);
-  }
-
+    Mantid::API::MatrixWorkspace &outputWorkspace, unsigned int specNum) {
   auto &inSpaceY = inputWorkspace.y(specNum);
+  const auto ySize = inSpaceY.size();
+
+  // Remove workspace indices which are out of index range for y space
+  removeElementsIf(workspaceIndices, [&ySize](const std::size_t &index) {
```

Comment:
This isn't a simple refactor, the actual behavior of how many values that are removed from the Y axis of the spectrum has changed. Now I understand that this will potentially catch any failures with data that is  and make it simpler to interact with, I however disagree with this approach. This is because this will remove potentially valid data and "do the sort", this algorithm should just fail in this case if it is a histogram. The algorithm is not stated to fix your broken data but just sort a workspace based on it's X axis.

If at this stage it needs to remove more than one workspace index from the passed vector, then something has gone really wrong with expected data and we should not process it. I would like to suggest reverting to only remove one value at the end of the vector given it is a histogram using previously generated bool values (This could be made a member variable) to keep previous efficiency. Further more I would like to suggest we throw if it should need to remove more than one, as we do not support that data in this algorithm, so it should not be processed, perhaps that could be done early in exec.

(Let alone the fact this is ran for every single data value passed to this function, which never had to occur for the original purpose of this algorithm which was actually sorting point data, not flipping a histograms X Axis.)

Reply:
I have added a check which will throw if the input workspace is not point data or a histogram workspace so that it is now safe to remove just the last workspaceIndex